File: challenges/295/python/solution.py
#!/usr/bin/env python
from typing import List

from challenges.utils import expect
from sortedcontainers import SortedList


class MedianFinderSimple:
    """
    A simple solution using SortedList container
    """

    def __init__(self):
        self.sorted_numbers = SortedList()

    def addNum(self, num: int) -> None:
        self.sorted_numbers.add(num)

    def findMedian(self) -> float:
        l = len(self.sorted_numbers)
        if l % 2 == 0:
            return (self.sorted_numbers[l // 2 - 1] + self.sorted_numbers[l // 2]) / 2
        return self.sorted_numbers[l // 2]


class MedianFinder:
    """
    Implemented using binary search. It's slower than the above but takes less memory.
    """

    ordered_numbers: List[int]

    def __init__(self):
        self.ordered_numbers = []

    def find_index(self, number):
        left, right = 0, len(self.ordered_numbers) - 1
        while left <= right:
            if self.ordered_numbers[left] >= number:
                return left
            if self.ordered_numbers[right] <= number:
                return right + 1

            middle_index = (right + left) // 2  # 1 => 0; 2 => 0; 3 => 1
            if number == self.ordered_numbers[middle_index]:
                return middle_index

            # [1, 5, 10]; 7 => 2
            # [1, 5, 10]; 0 => 0
            if self.ordered_numbers[middle_index] < number:
                left = middle_index + 1
            else:
                right = middle_index - 1

    def addNum(self, num: int) -> None:
        if not self.ordered_numbers:
            self.ordered_numbers = [num]
        else:
            self.ordered_numbers.insert(self.find_index(num), num)

    def findMedian(self) -> int | float | None:
        if not self.ordered_numbers:
            return None

        numbers_length = len(self.ordered_numbers)
        index = (numbers_length - 1) // 2
        if numbers_length % 2 == 1:
            return self.ordered_numbers[index]
        else:
            return (self.ordered_numbers[index] + self.ordered_numbers[index + 1]) / 2


if __name__ == "__main__":
    obj = MedianFinder()

    for method, param, result in zip(
        ["MedianFinder", "addNum", "addNum", "findMedian", "addNum", "findMedian"],
        [[], [1], [2], [], [3], []],
        [None, None, None, 1.50000, None, 2.00000],
    ):
        if method == "MedianFinder":
            continue
        expect(getattr(obj, method)(*param), result)

File: challenges/295/js/solution.ts
// has to be var to be compatible with LeetCode
import { expect, zip } from "../../util";

class MedianFinder {
  sortedNums: number[];

  constructor() {
    this.sortedNums = [];
  }

  findIndex(num: number): number {
    let [left, right] = [0, this.sortedNums.length - 1];

    while (left <= right) {
      if (this.sortedNums[left] >= num) {
        return left;
      }
      if (this.sortedNums[right] <= num) {
        return right + 1;
      }

      const middleIndex = Math.floor((left + right) / 2);
      const middleNum = this.sortedNums[middleIndex];
      if (middleNum === num) {
        return middleIndex;
      }

      //[2, 5, 10]; 3
      if (middleNum > num) {
        right = middleIndex - 1;
      } else {
        left = middleIndex + 1;
      }
    }

    return 0;
  }

  findMedian(): number | null {
    const length = this.sortedNums.length;
    if (!length) {
      return null;
    }

    const medianIndex = Math.floor((length - 1) / 2);
    if (length % 2 === 0) {
      return (
        (this.sortedNums[medianIndex] + this.sortedNums[medianIndex + 1]) / 2
      );
    } else {
      return this.sortedNums[medianIndex];
    }
  }

  addNum(num: number): null {
    if (!this.sortedNums.length) {
      this.sortedNums = [num];
    } else {
      this.sortedNums.splice(this.findIndex(num), 0, num);
    }
    return null;
  }
}

const obj: { [key: string]: any } = new MedianFinder();
for (const [method, params, output] of zip([
  ["MedianFinder", "addNum", "addNum", "findMedian", "addNum", "findMedian"],
  [[], [1], [2], [], [3], []],
  [null, null, null, 1.5, null, 2.0],
])) {
  if (method === "MedianFinder") {
    continue;
  }
  expect(obj[method as string](...params), output);
}
